File: residential_controller.js
//START   /* START THE ELEVATOR SYSTEM */
function startElevatorSystem(nbrOfFloors, nbrOfElevator) {
  var controller = new ElevatorController(nbrOfFloors, nbrOfElevator);
  return controller;
}
class Column {
  constructor(nbrOfFloors, nbrOfElevator) {
    this.nbrOfFloors = nbrOfFloors;
    this.nbrOfElevator = nbrOfElevator;
    this.elevatorList = [];
    for (let i = 0; i < this.nbrOfElevator; i++) {
      let elevator = new Elevator(i, "idle", 1, "UP");
      this.elevatorList.push(elevator);
    }
  }
}

class Elevator {
  constructor(elevatorId, status, currentFloor, currentDirection) {
    this.elevatorId = elevatorId;
    this.status = status;
    this.currentFloor = currentFloor;
    this.currentDirection = currentDirection;
    this.floorList = [];
  }
  //END     /* START THE ELEVATOR SYSTEM */


  //START     /* SEND THE REQUESTED FLOOR TO THE COMPUTE LIST */              
  sendRequest(RequestedFloor) {
    this.floorList.push(RequestedFloor);
    this.compute_list();
    this.operate_elevator(RequestedFloor);
  }

  compute_list() {
    if (this.currentDirection === "UP") {
      this.floorList.sort();
    } else if (this.currentDirection === "DOWN") {
      this.floorList.sort();
      this.floorList.reverse();
    }
    return this.floorList;
  }
  //END     /* SEND THE REQUESTED FLOOR TO THE COMPUTE LIST */              



  //START     /* SEND THE REQUESTED FLOOR TO THE OPERATE SYSTEM AND TAKE ACTION */                            
  operate_elevator(RequestedFloor) {
    while (this.floorList > 0) {
      if (RequestedFloor === this.currentFloor) {
        this.openDoor();
        this.status = "MOVING ... ";
        this.floorList.shift();
      } else if (RequestedFloor < this.currentFloor) {
        this.status = "MOVING ... ";
        console.log(">>>>>>>>>>>>>>>>>>>>>>>>");
        console.log("ELEVATOR " + this.elevatorId, this.status);
        console.log(">>>>>>>>>>>>>>>>>>>>>>>>");
        this.moveDown(RequestedFloor);
        this.status = "STOP";
        console.log(">>>>>>>>>>>>>>>>>>>>>>>>");
        console.log("ELEVATOR " + this.elevatorId, this.status);
        console.log(">>>>>>>>>>>>>>>>>>>>>>>>");
        this.openDoor();
        this.floorList.shift();
      } else if (RequestedFloor > this.currentFloor) {
        delay(1000);
        this.status = "MOVING ... ";
        console.log(">>>>>>>>>>>>>>>>>>>>>>>>");
        console.log("ELEVATOR " + this.elevatorId, this.status);
        console.log(">>>>>>>>>>>>>>>>>>>>>>>>");
        this.moveUp(RequestedFloor);
        this.status = "STOP";
        console.log(">>>>>>>>>>>>>>>>>>>>>>>>");
        console.log("ELEVATOR " + this.elevatorId, this.status);
        console.log(">>>>>>>>>>>>>>>>>>>>>>>>");

        this.openDoor();

        this.floorList.shift();
      }
    }
    if (this.floorList === 0) {
      this.status = "idle";
    }
  }
  requestFloorButton(RequestedFloor) {
    this.RequestedFloor = RequestedFloor;
    this.floorLight = floorLight;
  }
  callFloorButton(FloorNumber, Direction) {
    this.FloorNumber = FloorNumber;
    this.Direction = Direction;
  }
  //END     /* SEND THE REQUESTED FLOOR TO THE OPERATE SYSTEM AND TAKE ACTION */                            


  //START     /* OPEN AND CLOSE DOORS BUTTONS */              
  openDoor() {
    delay(1000);
    console.log("<-- OPEN DOOR -->");
    console.log(">>>>>>>>>>>>>>>>>>>>>>>>");
    console.log("BUTTON LIGHT OFF");
    delay(1000);

    console.log(">>>>>>>>>>>>>>>>>>>>>>>>");
    delay(1000);
    this.closeDoor();
  }
  closeDoor() {
    console.log("--> CLOSE DOOR <--");
    delay(1000);
  }
  //END     /* OPEN AND CLOSE DOORS BUTTONS */              


  //START     /* MOVE THE ELEVATOR UP */                            
  moveUp(RequestedFloor) {
    console.log("FLOOR : " + this.currentFloor);
    delay(1000);
    while (this.currentFloor !== RequestedFloor) {
      this.currentFloor += 1;
      console.log("FLOOR : " + this.currentFloor);

      delay(1000);
    }
  }
  //END     /* MOVE THE ELEVATOR UP */                            


  //START     /* MOVE THE ELEVATOR DOWN */                            
  moveDown(RequestedFloor) {
    console.log("FLOOR : " + this.currentFloor);
    delay(1000);
    while (this.currentFloor !== RequestedFloor) {
      this.currentFloor -= 1;
      console.log("FLOOR : " + this.currentFloor);

      delay(1000);
    }
  }
  //END     /* MOVE THE ELEVATOR DOWN */                            
}

class ElevatorController {
  constructor(nbrOfFloors, nbrOfElevator) {
    this.nbrOfFloors = nbrOfFloors;
    this.nbrOfElevator = nbrOfElevator;
    this.column = new Column(nbrOfFloors, nbrOfElevator);
    // console.log(this.column);

    console.log("START TEST");
  }

  //START     /* REQUEST AN ELEVATOR */                            
  RequestElevator(FloorNumber, Direction) {
    delay(1000);
    console.log(">>>>>>>>>>>>>>>>>>>>>>>>");
    console.log("REQUEST ELEVATOR TO THE FLOOR : ", FloorNumber);
    delay(1000);
    console.log(">>>>>>>>>>>>>>>>>>>>>>>>");
    console.log("CALL BUTTON LIGHT ON");
    delay(1000);

    let elevator = this.findBestElevator(FloorNumber, Direction);
    elevator.sendRequest(FloorNumber);
    return elevator;
  }
  //START   REQUEST AN ELEVATOR

  //START     /* REQUESTING THE FLOOR */                            
  RequestFloor(elevator, RequestedFloor) {
    delay(1000);
    console.log(">>>>>>>>>>>>>>>>>>>>>>>>");
    console.log("REQUESTED FLOOR : ", RequestedFloor);
    delay(1000);
    console.log(">>>>>>>>>>>>>>>>>>>>>>>>");
    console.log("BUTTON LIGHT ON");
    delay(1000);
    elevator.sendRequest(RequestedFloor);
  }
  //END     /* REQUESTING THE FLOOR */                            


  //START     /* LOOKING FOR THE BEST ELEVATOR */                            
  findBestElevator(FloorNumber, Direction) {
    console.log("LOOKING FOR THE BEST ELEVATOR", FloorNumber, Direction);

    let bestElevator = null;
    let bestDistance = 1000;
    for (let i = 0; i < this.column.elevatorList.length; i++) {
      let elevator = this.column.elevatorList[i];
      if (FloorNumber === elevator.currentFloor && (elevator.status === "STOP" || elevator.status === "idle" || elevator.status === "MOVING ... ")) {
        return elevator;
      } else {
        let referenceDistance = Math.abs(FloorNumber - elevator.currentFloor);
        if (bestDistance > referenceDistance) {
          bestDistance = referenceDistance;
          bestElevator = elevator;

          if (elevator.Direction === Direction) {
            bestElevator = elevator;
          }
        }
      }
    }
    return bestElevator;
    //END     /* LOOKING FOR THE BEST ELEVATOR */                            
  }
}
//START   REQUEST AN ELEVATOR


//START     /* DELAY TIME */   //https://stackoverflow.com/questions/1183872/put-a-delay-in-javascript                          
function delay(milliseconds) {
  var start = new Date().getTime();
  for (var i = 0; i < 1e7; i++) {
    if ((new Date().getTime() - start) > milliseconds) {
      break;
    }
  }
}
//END     /* DELAY TIME */   //https://stackoverflow.com/questions/1183872/put-a-delay-in-javascript                          


/*--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------*/
/* START THE ELEVATOR TEST */
function executeTest() {
  var controller = startElevatorSystem(1, 5);

  controller.column.elevatorList[0].currentFloor = 1;
  controller.column.elevatorList[0].status = "MOVING ... ";
  controller.column.elevatorList[0].currentDirection = "DOWN";
  controller.column.elevatorList[1].currentFloor = 5;
  controller.column.elevatorList[1].status = "MOVING ...";
  controller.column.elevatorList[1].currentDirection = "DOWN";



  var elevator = controller.RequestElevator(7, "UP");
  controller.RequestFloor(elevator, 6);
  elevator = controller.RequestElevator(3, "DOWN");
  controller.RequestFloor(elevator, 5);
  console.log("TEST FINISHED");
}
//END /* START THE ELEVATOR TEST */

File: residential_controller.py
#START THE ELEVATOR SYSTEM */
import time

class elevatorController:
    def __init__(self, nbrOfFloors, nbrOfElevator):
        self.nbrOfFloors = nbrOfFloors
        self.nbrOfElevator = nbrOfElevator
        self.column = Column(nbrOfFloors, nbrOfElevator)
        print("START TEST")
#END START THE ELEVATOR SYSTEM */


#START     /* REQUEST AN ELEVATOR */                            
    def RequestElevator(self, FloorNumber, Direction):
        time.sleep(1)
        print(">>>>>>>>>>>>>>>>>>>>>>>>")
        print("REQUEST ELEVATOR TO THE FLOOR : ", FloorNumber)
        time.sleep(1)
        print(">>>>>>>>>>>>>>>>>>>>>>>>")
        print("CALL BUTTON LIGHT ON")
        time.sleep(1)
        elevator = self.find_best_elevator(FloorNumber, Direction)
        elevator.send_request(FloorNumber)
        return elevator
#END     /* REQUEST AN ELEVATOR */                            

#START     /* REQUESTING FLOOR */ 
    def RequestFloor(self, elevator, RequestedFloor):
        time.sleep(1)
        print(">>>>>>>>>>>>>>>>>>>>>>>>")
        print("REQUESTED FLOOR : ", RequestedFloor)
        time.sleep(1)
        print(">>>>>>>>>>>>>>>>>>>>>>>>")
        print("BUTTON LIGHT ON")
        time.sleep(1)
        elevator.send_request(RequestedFloor)
#END     /* REQUESTING FLOOR */ 


#START     /* LOOKING FOR THE BEST ELEVATOR */                            
    def find_best_elevator(self, FloorNumber, Direction):
        bestElevator = None
        shortestDistance = 1000
        for elevator in (self.column.elevator_list):
            if (FloorNumber == elevator.elevatorFloor and (elevator.status == "STOP" or elevator.status == "idle" or elevator.status == " MOVING")):
                return elevator
            else:
                referenceDisrance = abs(FloorNumber - elevator.elevatorFloor)
                if shortestDistance > referenceDisrance:
                    shortestDistance = referenceDisrance
                    bestElevator = elevator

                elif elevator.Direction == Direction:
                    bestElevator = elevator

        return bestElevator


class Elevator:
    def __init__(self, elevatorNumb, status, elevatorFloor, elevator_direction):
        self.elevatorNumb = elevatorNumb
        self.status = status
        self.elevatorFloor = elevatorFloor
        self.elevator_direction = elevator_direction
        self.floor_list = []
#END     /* LOOKING FOR THE BEST ELEVATOR */                            

#START     /* SENDING A REQUEST */                            

    def send_request(self, RequestedFloor):
        self.floor_list.append(RequestedFloor)
        self.compute_list()
        self.operate_elevator(RequestedFloor)
#END     /* SENDING A REQUEST */                            

#START     /* COMPUTING LIST */                            
    def compute_list(self):
        if self.elevator_direction == "UP":
            self.floor_list.sort()
        elif self.elevator_direction == "DOWN":
            self.floor_list.sort()
            self.floor_list.reverse()
        return self.floor_list
#END     /* COMPUTING LIST */                            

#START     /* OPERATING THE ELEVATOR */                            
    def operate_elevator(self, RequestedFloor):
        while (len(self.floor_list) > 0):
            if ((RequestedFloor == self.elevatorFloor)):
                self.Open_door()
                self.status = " MOVING"

                self.floor_list.pop()
            elif (RequestedFloor < self.elevatorFloor):

                self.status = " MOVING"
                print(">>>>>>>>>>>>>>>>>>>>>>>>")
                print("Elevator", self.elevatorNumb, self.status)
                print(">>>>>>>>>>>>>>>>>>>>>>>>")
                self.Direction = "DOWN"
                self.Move_down(RequestedFloor)
                self.status = "STOP"
                print(">>>>>>>>>>>>>>>>>>>>>>>>")
                print("Elevator", self.elevatorNumb, self.status)
                print(">>>>>>>>>>>>>>>>>>>>>>>>")
                self.Open_door()
                self.floor_list.pop()
            elif (RequestedFloor > self.elevatorFloor):

                time.sleep(1)
                self.status = " MOVING"
                print(">>>>>>>>>>>>>>>>>>>>>>>>")
                print("Elevator", self.elevatorNumb, self.status)
                print(">>>>>>>>>>>>>>>>>>>>>>>>")
                self.Direction = "UP"
                self.Move_up(RequestedFloor)
                self.status = "STOP"
                print(">>>>>>>>>>>>>>>>>>>>>>>>")
                print("Elevator", self.elevatorNumb, self.status)
                print(">>>>>>>>>>>>>>>>>>>>>>>>")

                self.Open_door()

                self.floor_list.pop()

        if self.floor_list == 0:
            self.status = "idle"
#END     /* OPERATING THE ELEVATOR */ 

#START     /* OPEN AND CLOSE DOORS BUTTONS */              
    def Open_door(self):
        time.sleep(1)
        print("OPENING THE DOOR")
        print(">>>>>>>>>>>>>>>>>>>>>>>>")
        print("BUTTON LIGHT OFF")
        time.sleep(1)
        print(">>>>>>>>>>>>>>>>>>>>>>>>")
        time.sleep(1)
        self.Close_door()

    def Close_door(self):
        print("CLOSING THE DOORS")
        time.sleep(1)
#END     /* OPEN AND CLOSE DOORS BUTTONS */              


#START     /* MOOVING THE ELEVATOR UP OR DOWN */
    def Move_up(self, RequestedFloor):
        print("FLOOR : ", self.elevatorFloor)
        time.sleep(1)
        while(self.elevatorFloor != RequestedFloor):
            self.elevatorFloor += 1
            print("FLOOR : ", self.elevatorFloor)
            time.sleep(1)

    def Move_down(self, RequestedFloor):
        print("FLOOR : ", self.elevatorFloor)
        time.sleep(1)
        while(self.elevatorFloor != RequestedFloor):
            self.elevatorFloor -= 1
            print("FLOOR : ", self.elevatorFloor)
            
            time.sleep(1)
#START     /* MOOVING THE ELEVATOR UP OR DOWN */                            


#START     /* CALL BUTTON */                            
class callButton:
    def __init__(self, FloorNumber, Direction):
        self.FloorNumber = FloorNumber
        self.Direction = Direction
        self.light = False
#END     /* CALLING BUTTON */                            


#START     /* FLOOR BUTTON */
class floorButton:
    def __init__(self, RequestedFloor):
        self.RequestedFloor = RequestedFloor
#START     /* FLOOR  BUTTON */


class Column:
    def __init__(self, nbrOfFloors, nbrOfElevator):
        self.nbrOfFloors = nbrOfFloors
        self.nbrOfElevator = nbrOfElevator
        self.elevator_list = []
        for i in range(nbrOfElevator):
            elevator = Elevator(i, "idle", 1, "UP")
            self.elevator_list.append(elevator)


#----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
# START THE ELEVATOR TEST */

controller = elevatorController(10, 2)
controller.column.elevator_list[0].elevatorFloor = 2
controller.column.elevator_list[0].status = " MOVING"
controller.column.elevator_list[0].elevator_direction = "DOWN"
controller.column.elevator_list[1].elevatorFloor = 6
controller.column.elevator_list[1].status = " MOVING"
controller.column.elevator_list[1].elevator_direction = "DOWN"

elevator = controller.RequestElevator(9, "UP")
controller.RequestFloor(elevator, 3)
print("TEST FINISHED")
# END THE ELEVATOR TEST */

